set.mm

```diff
@@ -757432,6 +757435,319 @@ last digit of the integer part (for ` b = ` 10 the first digit before
   $}
 
 
+$(
+-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-
+  N-ary functions
+-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-
+
+  According to Wikipedia ("Arity", ~ https://en.wikipedia.org/wiki/Arity ,
+  19-May-2024): "In logic, mathematics, and computer science, _arity_ is the
+  number of arguments or operands taken by a function, operation or relation."
+
+  N-ary functions are often also called multivariate functions, without
+  indicating the actual number of argumens. See also Wikipedia
+  ("Multivariate functions",   19-May-2024,
+  ~ https://en.wikipedia.org/wiki/Function_(mathematics)#Multivariate_functions
+  ): "A _multivariate function_, multivariable function, or function of several
+  variables is a function that depends on several arguments.  ...  Formally, a
+  function of n variables is a function whose domain is a set of n-tuples.
+  For example, multiplication of integers is a function of two variables, or
+  _bivariate function_, whose domain is the set of all ordered pairs (2-tuples)
+  of integers, and whose codomain is the set of integers.  The same is true for
+  every binary operation.  Commonly, an n-tuple is denoted enclosed between
+  parentheses, such as in ( 1 , 2 , ... , n ).  When using functional notation,
+  one usually omits the parentheses surrounding tuples, writing
+  f ( x_1 , ... , x_n ) instead of f ( ( x_1 , ... , x_n ) ).  Given n sets
+  X_1 , ... , X_n , the set of all n-tuples ( x_1 , ... , x_n ) such that
+  x_1 is element of X_1 , ... , x_n is element of X_n is called the Cartesian
+  product of X_1 , ... , X_n , and denoted X_1 X ... X X_n .
+  Therefore, a multivariate function is a function that has a Cartesian product
+  or a proper subset of a Cartesian product as a domain: ` f : U --> Y ` where
+  where the domain ` U ` has the form
+  ` U C_ ( ( ... ( ( X `` 1 ) X. ( X `` 2 ) ) X. ... ) X. ( X `` n ) ) `."
+
+  In the following, n-ary functions are defined as mappings (see ~ df-map )
+  from a finite sequence of arguments, which themselves are defined as
+  mappings from the half-open range of nonnegative integers to the domain of
+  each argument.  Furthermore, the definition is restricted to endofunctions,
```

Comment:
I'm afraid that one not-so-distant day we might need a N-ary function with different range and domains (so, not endofunctions).

Actually the most generic definition would have each X different, as in [multivariate functions](https://en.wikipedia.org/wiki/Function_(mathematics)#Multivariate_functions).
We do have a definition of the [cartesian product of a set of sets](https://us.metamath.org/mpeuni/df-ixp.html) which could be of use.

Of course for this definition you have the primitive recursive functions in mind, so endofunctions are sufficient, and this development is already complex.

Reply:
I have chosen this approach to keep the representation simple (only two parameters `X`and `N` in `( X NaryF N )`. If the range had to be different, then we must add a third parameter, e.g. `( <. X , Y >. NaryF N )`.  If the domains for the arguments had to be arbitrary, the notation would become even more complex...

Reply:
Got it. This already has a wide range of applications.

Reply:
I have added ~ naryfvalixp which shows that n-ary (endo)functions can also be expressed with the notation of infinite Cartesian products. Nevertheless, I think using the notation of infinite Cartesian products to allow different domains for the arguments is not so easy, maybe as `NaryF = ( x e. _V , n e. NN0 |-> ( x ^m X_ y e. ( 0 ..^ n ) x) )`, because ` x ` must not depend on ` y ` for the soundness of the definition...